gstools/normalizer/base.py

```diff
@@ -0,0 +1,257 @@
+    """:class:`tuple`: Valid range for output/normal data."""
+
+    def __init__(self, data=None, **parameter):
+        # only use parameter, that have a provided default value
+        for key, value in self.default_parameter.items():
+            setattr(self, key, parameter.get(key, value))
+        # fit parameters if data is given
+        if data is not None:
+            self.fit(data)
+        # optimization results
+        self._opti = None
+        # precision for printing
+        self._prec = 3
+
+    def _denormalize(self, data):
+        return data
+
+    def _normalize(self, data):
+        return data
+
+    def _derivative(self, data):
+        return spm.derivative(self._normalize, data, dx=1e-6)
+
+    def _loglikelihood(self, data):
+        add = -0.5 * np.size(data) * (np.log(2 * np.pi) + 1)
+        return self._kernel_loglikelihood(data) + add
+
+    def _kernel_loglikelihood(self, data):
+        res = -0.5 * np.size(data) * np.log(np.var(self._normalize(data)))
+        return res + np.sum(np.log(np.maximum(1e-16, self._derivative(data))))
+
+    def _check_input(self, data, data_range=None, return_output_template=True):
+        is_data = np.array(np.logical_not(np.isnan(data)))
+        if return_output_template:
+            out = np.full_like(data, np.nan, dtype=np.double)
+        data = np.array(data, dtype=np.double)[is_data]
+        if data_range is not None and np.min(np.abs(data_range)) < np.inf:
+            dat_in = np.logical_and(data > data_range[0], data < data_range[1])
+            if not np.all(dat_in):
+                warnings.warn(
+                    "{0}: data (min: {1}, max: {2}) out of range: {3}. "
+                    "Affected values will be treated as NaN.".format(
+                        self.name, np.min(data), np.max(data), data_range
+                    )
+                )
+                is_data[is_data] &= dat_in
+                data = data[dat_in]
+        if return_output_template:
+            return data, is_data, out
+        return data
+
+    def denormalize(self, data):
+        """Transform to input distribution.
+
+        Parameters
+        ----------
+        data : array_like
+            Input data (normal distributed).
+
+        Returns
+        -------
+        :class:`numpy.ndarray`
+            Denormalized data.
+        """
+        data, is_data, out = self._check_input(data, self.denormalize_range)
+        out[is_data] = self._denormalize(data)
+        return out
+
+    def normalize(self, data):
+        """Transform to normal distribution.
+
+        Parameters
+        ----------
+        data : array_like
+            Input data (not normal distributed).
+
+        Returns
+        -------
+        :class:`numpy.ndarray`
+            Normalized data.
+        """
+        data, is_data, out = self._check_input(data, self.normalize_range)
+        out[is_data] = self._normalize(data)
+        return out
+
+    def derivative(self, data):
```

Comment:
Should the spacing `dx` maybe be included as an optional argument for the scipy-method?
Same than also for `likelihood`, ...

Reply:
the derivative is mostly overwritten to use an analytical solution. I will add a private `_dx` that could be accessed by users who really want to.